### PR TITLE
odf to odf: start using LoadBalancer type for ocs-provider-server svc

### PR DESCRIPTION
**Pull request by iamniting**
We were using service of type NodePort which is prune to failures
whenever a used node goes down or removed from a OCP cluster.

Start using the service of type LoadBalancer which creates a external IP
which can be used via consumers.

Bug: https://bugzilla.redhat.com/show_bug.cgi?id=2060487

Signed-off-by: Nitin Goyal <nigoyal@redhat.com>

### CODE REVIEW

File: controllers/storagecluster/provider_server.go [hunk 1]
```diff
@@ -26,8 +25,7 @@ const (
 	ocsProviderServerName  = "ocs-provider-server"
 	providerAPIServerImage = "PROVIDER_API_SERVER_IMAGE"
 
-	ocsProviderServicePort     = int32(50051)
-	ocsProviderServiceNodePort = int32(31659)
+	ocsProviderServicePort = int32(50051)
```
Comment (oritwas) [inline controllers/storagecluster/provider_server.go:28]: Will it be possible to make the port configurable i.e part of the CRD?

Comment (Madhu-1) [reply controllers/storagecluster/provider_server.go:28]: If I understand correctly @oritwas is asking Can this port be configurable like the user can provide a custom port when deploying OCS in provider mode? like in CSV/CRD or POD ENV etc?

Comment (Madhu-1) [reply controllers/storagecluster/provider_server.go:28]: not sure POD ENV is useful

Comment (iamniting) [reply controllers/storagecluster/provider_server.go:28]: We can read it from the storagecluster CR if required. I am not in favor of the ENV variables as we have to populate that value either via CSV on the build time or via configmap.

File: controllers/storagecluster/provider_server.go [hunk 2]
```diff
@@ -175,19 +173,22 @@ func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance
...
+		err = fmt.Errorf("Service %s Status.LoadBalancer.Ingress is empty", actualService.Name)
+		r.Log.Error(err, "Service status is not set yet", "Status", actualService.Status)
 		return err
 	}
 
-	if len(nodeAddresses) == 0 {
-		err = fmt.Errorf("Did not found any worker node addresses")
-		r.Log.Error(err, "Worker nodes count is zero")
+	if actualService.Status.LoadBalancer.Ingress[0].IP != "" {
+		instance.Status.StorageProviderEndpoint = actualService.Status.LoadBalancer.Ingress[0].IP + ":" + fmt.Sprint(ocsProviderServicePort)
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:183]:
```suggestion
		instance.Status.StorageProviderEndpoint = fmt.Sprint(%s:$d", actualService.Status.LoadBalancer.Ingress[0].IP, ocsProviderServicePort)
```

File: controllers/storagecluster/provider_server.go [hunk 3]
```diff
@@ -175,19 +173,22 @@ func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance
...
+	} else if actualService.Status.LoadBalancer.Ingress[0].Hostname != "" {
+		instance.Status.StorageProviderEndpoint = actualService.Status.LoadBalancer.Ingress[0].Hostname + ":" + fmt.Sprint(ocsProviderServicePort)
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:185]: Same as above

File: controllers/storagecluster/provider_server.go (hunk 1 above)
Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:29]: Why complicate things? just have it as a constant. This is for provider mode which is always deployed on its own cluster which means that the chance of collision is almost 0. So why are trying to solve a non-issue

File: controllers/storagecluster/provider_server.go [hunk 4]
```diff
@@ -48,12 +47,17 @@ func (o *ocsProviderServer) ensureCreated(r *StorageClusterReconciler, instance
...
 		o.createSecret,
 		o.createService,
 		o.createDeployment,
 	} {
-		err := f(r, instance)
+		res, err := f(r, instance)
+
+		if !res.IsZero() {
+			return res, err
+		}
```
[the remark this answers is not in the document]
Comment (iamniting) [reply controllers/storagecluster/provider_server.go:59]: If there is an error only, we are not returning immediately. We are doing all 3 operations, collecting the errors, and returning them together.

File: controllers/storagecluster/provider_server.go [hunk 5]
```diff
@@ -168,32 +172,33 @@ func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance
...
 	}
 
 	r.Log.Info("Service create/update succeeded")
 
-	nodeAddresses, err := r.getWorkerNodesInternalIPAddresses()
-	if err != nil {
-		return err
+	if len(actualService.Status.LoadBalancer.Ingress) == 0 {
+		r.Log.Info("Waiting for Service Status.LoadBalancer.Ingress", "Service", actualService.Name, "Status", actualService.Status)
+		return reconcile.Result{RequeueAfter: time.Second * 5}, nil
```
[the remark this answers is not in the document]
Comment (iamniting) [reply controllers/storagecluster/provider_server.go]: done

File: controllers/storagecluster/provider_server.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (iamniting) [reply controllers/storagecluster/provider_server.go]: made changes to return result and error immediately

File: controllers/storagecluster/provider_server.go [hunk 6]
```diff
@@ -46,21 +45,19 @@ func (o *ocsProviderServer) ensureCreated(r *StorageClusterReconciler, instance
...
-			multierr.AppendInto(&finalErr, err)
-			continue
+		res, err := f(r, instance)
+
+		if !res.IsZero() || err != nil {
+			return res, err
 		}
 	}
 
-	return reconcile.Result{}, finalErr
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:63]: Why complicate things with a loop? 
You force every create func to have the exact same sig! no args and 2 return values. Why put these restrictions on the developer?

```suggestion
	if err := o.createSecret; err != nil {
		return reconcile.Result{}, err
	}
	
	if result, err: = o.createService; err != nil {
		return err
	} else if !res.IsZero() || {
		return result, nil
	} 
	
	if err := o.createDeployment; err != nil {
		return reconcile.Result{}, err
	}
	
	return reconcile.Result{}, nil
 ```

File: controllers/storagecluster/provider_server.go [hunk 7]
```diff
@@ -96,7 +93,7 @@ func (o *ocsProviderServer) ensureDeleted(r *StorageClusterReconciler, instance
 	return reconcile.Result{}, finalErr
 }
 
-func (o *ocsProviderServer) createDeployment(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+func (o *ocsProviderServer) createDeployment(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:96]: No need to introduce a result into the sig of this func. Please revert

File: controllers/storagecluster/provider_server.go [hunk 8]
```diff
@@ -168,32 +166,37 @@ func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance
...
+	instance.Status.StorageProviderEndpoint = endpoint
 
 	r.Log.Info("status.storageProviderEndpoint is updated", "Endpoint", instance.Status.StorageProviderEndpoint)
 
-	return nil
+	return reconcile.Result{}, nil
 }
 
-func (o *ocsProviderServer) createSecret(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+func (o *ocsProviderServer) createSecret(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:199]: No need to introduce a result into the sig of this func. Please revert

File: controllers/storagecluster/provider_server.go [hunk 9]
```diff
@@ -168,32 +166,37 @@ func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance
...
 	r.Log.Info("Service create/update succeeded")
 
-	nodeAddresses, err := r.getWorkerNodesInternalIPAddresses()
-	if err != nil {
-		return err
+	endpoint := ""
+
+	if len(actualService.Status.LoadBalancer.Ingress) != 0 {
+		if actualService.Status.LoadBalancer.Ingress[0].IP != "" {
+			endpoint = fmt.Sprintf("%s:%d", actualService.Status.LoadBalancer.Ingress[0].IP, ocsProviderServicePort)
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:180]:
```suggestion
			endpoint = actualService.Status.LoadBalancer.Ingress[0].IP
```

File: controllers/storagecluster/provider_server.go [hunk 10]
```diff
@@ -168,32 +166,37 @@ func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance
...
+		} else if actualService.Status.LoadBalancer.Ingress[0].Hostname != "" {
+			endpoint = fmt.Sprintf("%s:%d", actualService.Status.LoadBalancer.Ingress[0].Hostname, ocsProviderServicePort)
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:182]:
```suggestion
			endpoint = actualService.Status.LoadBalancer.Ingress[0].Hostname
```

File: controllers/storagecluster/provider_server.go [hunk 11]
```diff
@@ -168,32 +166,37 @@ func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance
...
-		r.Log.Error(err, "Worker nodes count is zero")
-		return err
+	if endpoint == "" {
+		r.recorder.ReportIfNotPresent(instance, corev1.EventTypeNormal, "Waiting", "Waiting for Ingress on service "+actualService.Name)
+		r.Log.Info("Waiting for Ingress on service", "Service", actualService.Name, "Status", actualService.Status)
+		return reconcile.Result{RequeueAfter: time.Second * 5}, nil
 	}
 
-	instance.Status.StorageProviderEndpoint = nodeAddresses[0] + ":" + fmt.Sprint(ocsProviderServiceNodePort)
+	instance.Status.StorageProviderEndpoint = endpoint
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:192]:
```suggestion
	instance.Status.StorageProviderEndpoint = fmt.Sprintf("%s:%d", endpoint, ocsProviderServicePort)
```

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/provider_server.go (hunk 6 above)
Comment (iamniting) [reply controllers/storagecluster/provider_server.go]: done

File: controllers/storagecluster/provider_server.go (hunk 10 above)
Comment (iamniting) [reply controllers/storagecluster/provider_server.go]: done

File: controllers/storagecluster/provider_server.go (hunk 7 above)
Comment (iamniting) [reply controllers/storagecluster/provider_server.go:100]: I changed this one as I am returning a requeue in this one as well when the server pod is not ready.

Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:100]: ok

**Pull request merged**